Lib/ufo2ft/filters/propagateAnchors.py

```diff
@@ -39,51 +39,61 @@ def __call__(self, font, glyphSet=None):
     def filter(self, glyph):
         if not glyph.components:
             return False
-        _propagate_glyph_anchors(glyph.font, glyph, self.context.processed)
+        _propagate_glyph_anchors(self.context.font, glyph,
+                                 self.context.processed)
         return True
 
 
-def _propagate_glyph_anchors(ufo, parent, processed):
-    """Propagate anchors for a single parent glyph."""
+def _propagate_glyph_anchors(font, composite, processed):
+    """
+    Propagate anchors from base glyphs to a given composite
+    glyph, and to all composite glyphs used in between.
+    """
 
-    if parent.name in processed:
+    if composite.name in processed:
+        return
+    processed.add(composite.name)
+
+    if not composite.components:
         return
-    processed.add(parent.name)
 
     base_components = []
     mark_components = []
     anchor_names = set()
     to_add = {}
-    for component in parent.components:
-        glyph = ufo[component.baseGlyph]
-        _propagate_glyph_anchors(ufo, glyph, processed)
+    for component in composite.components:
+        glyph = font[component.baseGlyph]
+        _propagate_glyph_anchors(font, glyph, processed)
         if any(a.name.startswith('_') for a in glyph.anchors):
             mark_components.append(component)
         else:
             base_components.append(component)
             anchor_names |= {a.name for a in glyph.anchors}
 
     for anchor_name in anchor_names:
-        # don't add if parent already contains this anchor OR any associated
-        # ligature anchors (e.g. "top_1, top_2" for "top")
-        if not any(a.name.startswith(anchor_name) for a in parent.anchors):
-            _get_anchor_data(to_add, ufo, base_components, anchor_name)
+        # don't add if composite glyph already contains this anchor OR any
+        # associated ligature anchors (e.g. "top_1, top_2" for "top")
+        if not any(a.name.startswith(anchor_name) for a in composite.anchors):
+            _get_anchor_data(to_add, font, base_components, anchor_name)
 
     for component in mark_components:
-        _adjust_anchors(to_add, ufo, component)
+        _adjust_anchors(to_add, font, component)
 
     # we sort propagated anchors to append in a deterministic order
     for name, (x, y) in sorted(to_add.items()):
         anchor_dict = {'name': name, 'x': x, 'y': y}
-        parent.appendAnchor(glyph.anchorClass(anchorDict=anchor_dict))
+        try:
+            composite.appendAnchor(anchor_dict)
+        except TypeError:
```

Comment:
we don't test against fontParts anyway, that line will not be hit..
I wasn't suggesting we should specifically support that. My point was more about simplifying that line, and avoid using too defconian methods which we may not want to keep in the future `fontTools.ufoLib`.
A `# pragma: no cover` should make sure this branch is not taken into account for test coverage

Reply:
I added `# pragma: no cover`.